Sources/PubNub/Models/PubNubPage.swift

```diff
@@ -0,0 +1,208 @@
+  var limit: Int? { get }
+
+  /// Allows for converting  between different PubNubBoundedPage types
+  init(from other: PubNubBoundedPage) throws
+}
+
+extension PubNubBoundedPage {
+  /// Converts this protocol into a custom type
+  /// - Parameter into: The explicit type for the returned value
+  /// - Returns: The protocol intiailized as a custom type
+  /// - Throws: An error why the custom type was unable to be created using this protocol instance
+  public func transcode<T: PubNubBoundedPage>(into _: T.Type) throws -> T {
+    return try transcode()
+  }
+
+  /// Converts this protocol into a custom type
+  /// - Returns: The protocol intiailized as a custom type
+  /// - Throws: An error why the custom type was unable to be created using this protocol instance
+  public func transcode<T: PubNubBoundedPage>() throws -> T {
+    // Check if we're already that object, and return
+    if let custom = self as? T {
+      return custom
+    }
+
+    return try T(from: self)
+  }
+}
+
+// MARK: Concrete Base Class
+
+/// The default implementation of the `PubNubBoundedPage` protocol
+public struct PubNubBoundedPageBase: PubNubBoundedPage, Codable, Hashable {
+  public let start: Timetoken?
+  public let end: Timetoken?
+  public let limit: Int?
+
+  public init?(start: Timetoken? = nil, end: Timetoken? = nil, limit: Int? = nil) {
+    if start == nil, end == nil, limit == nil {
+      return nil
+    }
+
+    self.start = start
+    self.end = end
+    self.limit = limit
+  }
+
+  public init(from other: PubNubBoundedPage) throws {
+    start = other.start
+    end = other.end
+    limit = other.limit
+  }
+}
+
+// MARK: - Hashed Page
+
+/// A cursor for the next page of a remote data
+public protocol PubNubHashedPage {
+  /// The hash value representing the netxt set of data
+  var start: String? { get }
+  /// The hash value representing the previous set of data
+  var end: String? { get }
+  /// The total count of all objects withing range
+  var totalCount: Int? { get }
+
+  /// Allows for converting  between different PubNubHashedPage types
+  init(from other: PubNubHashedPage) throws
+}
+
+extension PubNubHashedPage {
+  /// Converts this protocol into a custom type
+  /// - Parameter into: The explicit type for the returned value
+  /// - Returns: The protocol intiailized as a custom type
+  /// - Throws: An error why the custom type was unable to be created using this protocol instance
+  public func transcode<T: PubNubHashedPage>(into _: T.Type) throws -> T {
+    return try transcode()
+  }
+
+  /// Converts this protocol into a custom type
+  /// - Returns: The protocol intiailized as a custom type
+  /// - Throws: An error why the custom type was unable to be created using this protocol instance
+  public func transcode<T: PubNubHashedPage>() throws -> T {
+    // Check if we're already that object, and return
+    if let custom = self as? T {
+      return custom
+    }
+
+    return try T(from: self)
+  }
+}
+
+
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5691345)